### PR TITLE
Do SJA1105 hardware handling in kernel driver module instead of using spidev

### PR DESCRIPTION
**Pull request by gwaibel**
Following features are implemented:

* Probe the sja1105 chip
* Load a static configuration from user space (firmware_request).
* The configuration is loaded from the user space at driver probe and on request from the user space via sysfs.
* Register a network device for each port of the switch to handle the link status of the connected PHY.
* Dynamically adjust a port's MAC speed on link status change.
* Provide statistics information of a port to the net device interface.
* Provide a sysfs attribute to read general und port specific status information.
* Provide a sysfs attribute to reset the port status counters.
* Provide a sysfs attribute to access registers of the chip (raw register access)
* Provide a sysfs attribute to get the mapping between a net device and the switch's port number.
* Provide a sysfs attribute to upload a new static configuration to the chip.

See docs/md/sja1105-kmod.1.md for more details...

### CODE REVIEW

**Comment by vladimiroltean** [comment]
Updated the implementation with the following:
* Support for port counters and link speed control in ethtool
* Support for the PTP clock manipulation (using e.g. `phc2sys -s eth0 -c sethi -O 0 -m -l 7 -P 0.4 -I 0.01 -S 0.001`)
* Support for synchronizing the Time Aware Scheduler to the PTP clock
* Support for automatic management route setup in the port `ndo_xmit` callback (towards destination MAC addresses configured in MAC_FLT)

**Comment by sylane** [comment]
> Updated the implementation with the following:
> 
>     * Support for port counters and link speed control in ethtool
> 
>     * Support for the PTP clock manipulation (using e.g. `phc2sys -s eth0 -c sethi -O 0 -m -l 7 -P 0.4 -I 0.01 -S 0.001`)
> 
>     * Support for synchronizing the Time Aware Scheduler to the PTP clock
> 
>     * Support for automatic management route setup in the port `ndo_xmit` callback (towards destination MAC addresses configured in MAC_FLT)

Will synchronizing time with PTP over switched ports be supported ? What I mean, is that being able to have the boxes time-synchronized without requiring an extra network link to a non-switch port.

**Comment by vladimiroltean** [comment]
The switch ports will not participate actively in PTP synchronization over Ethernet (e.g. you still won't be able to run `ptp4l -i seth2`).
As I see the use case, you run PTP synchronization over the switch-facing Ethernet interface of the SoC (`ptp4l -i eth2`) and then you synchronize (over SPI, not over Ethernet) the switch clock to the eth2 clock (`phc2sys -s eth0 -c seth2`).
So this is indeed "over" the switch ports (as they carry PTP packets passively) but not synchronizing the switch ports per se.
If this matches what you need, then yes, we'll be making sure that this works properly. If not, we're open to suggestions.

**Comment by sylane** [comment]
I am not sure to completely understand what you describe just yet, but it sounds like it could work for us. As I added when editing my last message (I shouldn't have), in our project we could have a large mesh of routers (1000+) and they need to be synchronized, but we can't possibly have all of them wired with an extra Ethernet cable to some switch. So we must be able to have them synchronize over the switched network.

**Comment by vladimiroltean** [comment]
>  But then why eth0?

Mistake, I meant `eth2` (the switch-facing host port). I also corrected the original message.

> Would it be possible to have each routers in the mesh synchronize with there siblings, or all of them would need to use some centralized service ?

As far as I understand PTP, in the LAN an ad-hoc clock distribution tree gets formed. The root of the tree is called grandmaster (a type of ordinary clock), leaves are also ordinary clocks and the intermediary nodes are called boundary clocks. At each Ethernet link between 2 nodes, one of them is master and the other is slave, and the master tells the slave what time is it. The intermediary nodes (boundary clocks) have one port acting as slave and the others acting as master, and that is how PTP time is propagated. So PTP is centralized in the sense that there is a grandmaster, but decentralized in the sense that nodes synchronize to their directly-connected peers. Could you rephrase your question?

**Comment by sylane** [comment]
So the potential issue I can see is that we would require to have a PTP master have one of the switched port to be for its own master and some of the remaining port to be to slaves... We cannot really use eth2 as a whole because then it wouldn't be possible to form a tree.

**Comment by vladimiroltean** [comment]
So basically you want to form a boundary clock out of separate switch ports? I have to talk to a PTP expert to understand what can be done, but the issues I see are:
* The switch port driver needs to implement both explicit Rx and explicit Tx. Let's say that the "management routes" implementation will do as Tx substitute. But I don't know about explicit Rx on a switch port.
* The 5 switch ports share the same PTP clock. Not really sure whether this is an issue or not.

But as far as I know, a PTP master can be switched to multiple slaves with no issues. In fact that's what we do for the 3-board synchronized TAS demo, where we run `ptp4l -i eth0` on all 3 board and we connect them through a regular L2 switch. One of the 3 boards randomly becomes a master and the other 2 slaves.

**Comment by sylane** [comment]
If I understand correctly, using `eth2` is the same as using `eth0` if `eth0` is connected to a switch... So maybe there is a chance that it just works. The main issue will probably be that they are connected as a mesh network, not a tree like it is usual with normal switched network. If we have control on which port is for slave/master we can build a proper clock tree that would prevent loops, but without it is is probable that PTP will just get confused.

**Comment by sylane** [comment]
Am I correct in thinking that for now, `ptp4l` is updating the switch fabric clock automatically without the need of `phc2sys` when setting the parameter `-t N` ?

**Comment by vladimiroltean** [comment]
> for now

If this means "without pull request 36", then the answer is that ptp4l compiled with the SJA1105_ROOTDIR flag will have custom code that adjusts the PTP clock through libsja1105.
With pull request 36, compiling with SJA1105_ROOTDIR is no longer needed for this purpose, as the interaction is done through the /dev/ptp1 node, which the kmod driver creates.

**Comment by sylane** [comment]
So as it is right now with the `SJA1105_PTP_SYNC` patch, `ptp4l` is synchronizing the switch fabric properly, but there is no way to synchronize the system clock as `phc2sys` is missing. Is that correct ?

**Comment by vladimiroltean** [comment]
Not quite, no. The use case that we have is to synchronize the switch clock to the Ethernet port clock. That is not changing, only the way we do it is changing:
(before) using a patched ptp4l
(after) using phc2sys

**Comment by sylane** [comment]
ok, so `phc2sys` can be used to either synchronize the system clock or the switch clock that I wrongly though was the same as the ethernet clock. Right now both the ethernet port and switch clock would get synchronized but there is no way to synchronize the system clock due to the missing `phc2sys` dependency. Am I correct  ?

**Comment by vladimiroltean** [comment]
This has nothing to do with the system clock, it is just the program name that is confusing.
See `man phc2sys`:
```
NAME
       phc2sys - synchronize two or more clocks
```
This patchset only exposes the SJA1105 PHC the way it should be: as a `/dev/ptp1` node. Then you can interact with it using whatever standard tooling is available to you: `phc2sys`, `phc_ctl` etc, maybe even `ptp4l` under certain circumstances.

**Comment by sylane** [comment]
What I mean is that we are using `OpenIL-201808` and the command `phc2sys` is not available, so I cannot synchronize the system clock. Was the command added in `OpenIL-201810` or am I missing some configuration ?

**Comment by vladimiroltean** [comment]
Apply this patch on the OpenIL tree:
```
From d2653c3fe860c99ebf5cfc7c297e07d301663cc1 Mon Sep 17 00:00:00 2001
From: Vladimir Oltean <vladimir.oltean@nxp.com>
Date: Fri, 15 Jun 2018 13:27:58 +0300
Subject: [PATCH] package:ptp4l: ship all linux-ptp apps (hwstamp_ctl, phc_ctl,
 phc2sys, pmc)

Signed-off-by: Vladimir Oltean <vladimir.oltean@nxp.com>
---
 package/ptp4l/ptp4l.mk | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/package/ptp4l/ptp4l.mk b/package/ptp4l/ptp4l.mk
index 84ff6ed..57659fa 100644
--- a/package/ptp4l/ptp4l.mk
+++ b/package/ptp4l/ptp4l.mk
@@ -28,9 +28,9 @@ define PTP4L_BUILD_CMDS
 endef
 
 define PTP4L_INSTALL_TARGET_CMDS
-	$(INSTALL) -D -m 0755 $(@D)/ptp4l $(TARGET_DIR)/usr/sbin/ptp4l
+	$(foreach binary, ptp4l hwstamp_ctl phc_ctl phc2sys pmc,
+		$(INSTALL) -Dm 0755 $(@D)/$(binary) $(TARGET_DIR)/usr/sbin/$(binary))
 	cp -dpfr $(@D)/ptp4l_default.cfg $(TARGET_DIR)/etc/
-	cp -dpfr $(@D)/sja1105-ptp-tc.sh $(TARGET_DIR)/usr/sbin/sja1105-ptp-free-tc.sh
 endef
 
 $(eval $(generic-package))
-- 
2.7.4
```
Then rebuild as usual:
```
rm -rf output/build/ptp4l*
rm -rf dl/ptp4l*
make ptp4l
```

**Comment by sylane** [comment]
Thank you very much !

**Comment by vladimiroltean** [comment]
Obsoleted by upstream driver.
